x-pack/platform/test/security_api_integration/tests/login_selector/basic_functionality.ts

```diff
@@ -575,77 +609,197 @@ export default function ({ getService }: FtrProviderContext) {
+              .ca(CA_CERT)
+              .set('kbn-xsrf', 'xxx')
+              .set('Cookie', sharedCookie ? sharedCookie.cookieString() : '')
+              .send({
+                providerType: 'saml',
+                providerName: 'saml2',
+                currentURL: `https://kibana.com/login?next=/abc/xyz/${i}`,
+              })
+              .expect(200);
+
+            if (!sharedCookie) {
+              sharedCookie = parseCookie(samlHandshakeResponse.headers['set-cookie'][0])!;
+            }
+
+            const cookie = parseCookie(samlHandshakeResponse.headers['set-cookie'][0])!;
+            beforeCookies.push(cookie);
+
+            const samlRequestId = await getSAMLRequestId(samlHandshakeResponse.body.location);
+            const samlResponse = await createSAMLResponse({
+              issuer: `http://www.elastic.co/saml2`,
+              inResponseTo: samlRequestId,
+            });
 
-          const cookie = parseCookie(samlHandshakeResponse.headers['set-cookie'][0])!;
-          const samlRequestId = await getSAMLRequestId(samlHandshakeResponse.body.location);
-          const samlResponse = await createSAMLResponse({
-            issuer: `http://www.elastic.co/saml2`,
-            inResponseTo: samlRequestId,
+            samlResponseMapByRequestId[samlRequestId] = { samlResponse, cookie };
+          }
+          await es.indices.refresh({ index: '.kibana_security_session*' });
+
+          // Get all authenticated sessions from ES for verification
+          const sessionResponse = await es.search<SessionValue>({
+            index: '.kibana_security_session*',
+            size: 100,
+            query: {
+              bool: {
+                must: [
+                  {
+                    match_all: {},
+                  },
+                ],
+              },
+            },
           });
 
-          samlResponseMapByRequestId[samlRequestId] = { samlResponse, cookie };
-        }
+          const totalHits =
+            typeof sessionResponse.hits.total === 'number'
+              ? sessionResponse.hits.total
+              : sessionResponse.hits.total?.value;
+
+          // There should be only one intermediate session with all requestIds in it
+          expect(totalHits).to.equal(1);
+
+          const sources = sessionResponse.hits.hits.map((hit) => {
+            if (!hit._source) {
+              return { id: hit._id };
+            }
+            return {
+              id: hit._id,
+            };
+          });
 
-        const preparedCallbacks = [];
+          const { id: intermediateSessionId } = sources[0];
+
+          const preparedCallbacks = [];
+
+          let newCookie: Cookie | undefined;
+          let firstResponse: Response | undefined;
+
+          for (const requestId of Object.keys(samlResponseMapByRequestId)) {
+            const index = Object.keys(samlResponseMapByRequestId).indexOf(requestId);
+            const samlValues = samlResponseMapByRequestId[requestId];
```

Comment:
__nit:__ what about this?
```
...
samlResponseMapByRequestId[samlRequestId] = { samlResponse, cookie, index: i };
...

for (const { cookie, samlResponse, index } of Object.values(samlResponseMapByRequestId)) {
```

Also after such change, do we even need `samlRequestId` for anything? If not, we can use array instead of object to store responses, cookies and indices.